utils/src/qrcode.rs

```diff
@@ -0,0 +1,248 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub const FILLED_CHAR: char = '█';
+pub const EMPTY_CHAR: char = ' ';
+pub const NEW_LINE: char = '\n';
+
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum QrCodeError {
+    #[error("Given data is too long to fit in a QR code: {0}")]
+    DataTooLong(usize),
+    #[error("Border scaling failed. Possible very large border size: {0}")]
+    BorderScalingFailed(u32),
+}
+
+pub trait QrCode {
+    /// QR Codes are strictly squares, so this returns the side length
+    #[must_use]
+    fn side_length(&self) -> usize;
+
+    /// Returns the pixel at coordinate (x, y) if it exists, otherwise None
+    #[must_use]
+    fn pixel(&self, x: usize, y: usize) -> Option<bool>;
+
+    /// If the pixel is out of bounds, return false
+    #[must_use]
+    fn pixel_or_false(&self, x: usize, y: usize) -> bool {
+        self.pixel(x, y).unwrap_or(false)
+    }
+
+    /// Returns the QR code as a vector of booleans, where true represents a filled pixel and false
+    /// represents an empty pixel.
+    /// This contains all the information required to reconstruct the QR code. The side-length
+    /// can be calculated by taking the square root of the length of the vector
+    #[must_use]
+    fn as_vec(&self) -> Vec<bool> {
+        let mut result = Vec::with_capacity(self.side_length() * self.side_length());
+        for y in 0..self.side_length() {
+            for x in 0..self.side_length() {
+                result.push(self.pixel_or_false(x, y));
+            }
+        }
+        result
+    }
+
+    /// Returns a string representation of the QR code, using the given characters as placeholders
+    /// for empty, filled pixels, and the new line character
+    /// border_size is the number of pixels to add around the QR code
+    #[must_use]
+    fn encode_to_string(
+        &self,
+        border_size: u8,
+        empty_char: char,
+        filled_char: char,
+        new_line: char,
+    ) -> String {
+        let mut result = String::with_capacity(2 * self.side_length() * self.side_length());
+        let border: i32 = border_size as i32;
+        for y in -border..self.side_length() as i32 + border {
+            for x in -border..self.side_length() as i32 + border {
+                let c: char = if self.pixel_or_false(x as usize, y as usize) {
+                    filled_char
+                } else {
+                    empty_char
+                };
+                // We push twice because in a standard terminal, chars are rectangular,
+                // not square, with the height being twice the width (9x18 pixels).
+                // This makes the QR code pixels become squares.
+                result.push(c);
+                result.push(c);
```

Comment:
Maybe to make the displayed qr code a bit smaller, we can make use of the half-filled unicode chars "▀", "▄", and display two "pixels" per char vertically. That would be a fair bit more complex to implement though, so probably not worth the effort at the moment.

Reply:
Yep. That'd be nice. This suffices at the moment.